File: 0047. Permutations II/Permutations II.cpp
class Solution {
    void backtracking(vector<int> &nums, vector<int> &tmp, vector<vector<int>> &res, vector<bool> &visited) {
        if (tmp.size() == nums.size()) {
            res.push_back(tmp);
        } else {
            for (int i = 0; i < nums.size(); i++) {
                // visited[i - 1] is false means that the number is already used in the previous recursion
                if (visited[i] || i > 0 && nums[i] == nums[i - 1] && !visited[i - 1]) continue;
                visited[i] = true;
                tmp.push_back(nums[i]);
                backtracking(nums, tmp, res, visited);
                visited[i] = false;
                tmp.pop_back();
            }
        }
    }
    
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<int> tmp;
        vector<vector<int>> res;
        vector<bool> visited(nums.size(), false);
        backtracking(nums, tmp, res, visited);
        return res;
    }
};

File: 0047. Permutations II/Permutations II.py
class Solution:
    def backtracking(self, nums: List[int], tmp: List[int], res: List[List[int]], visited: List[int]) -> List[List[int]]:
        if len(nums) == len(tmp):
            res.append(tmp[:])
        else:
            for i in range(len(nums)):
                if visited[i] or i > 0 and nums[i - 1] == nums[i] and visited[i - 1] is False:
                    continue
                visited[i] = True
                tmp.append(nums[i])
                self.backtracking(nums, tmp, res, visited)
                visited[i] = False
                tmp.pop()
    
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res, tmp, visited = list(), list(), [False for i in range(len(nums))]
        self.backtracking(nums, tmp, res, visited)
        return res